src/pudl/helpers.py

```diff
@@ -339,6 +339,227 @@ def is_doi(doi):
+            cases if doing a left merge E.g. ``generation_eia923``.
+            Must contain columns specified by ``left_date_col`` and
+            ``shared_merge_cols`` argument.
+        right: The right dataframe in the merge. Typically annual in our uses
+            cases if doing a left merge E.g. ``generators_eia860``.
+            Must contain columns specified by ``right_date_col`` and and
+            ``shared_merge_cols`` argument.
+        left_date_col: Column in ``left`` containing datetime like data. Default is
+            ``report_date``. Must be convertible to a Datetime using
+            :func:`pandas.to_datetime`
+        new_date_col: Name of the reconstructed datetime column in the output dataframe.
+        right_date_col: Column in ``right`` containing datetime like data. Default is
+            ``report_date``. Must be convertible to a Datetime using
+            :func:`pandas.to_datetime`.
+        shared_merge_cols: The columns to merge on that are shared between both
+            dataframes. Typically ID columns like ``plant_id_eia``, ``generator_id``
+            or ``boiler_id``.
+        temporal_merge_cols: The temporal columns to merge on. Values must be
+            [``year``, ``quarter``, ``month``, ``day``]. E.g. if an annually reported
+            dataframe is being merged onto a monthly reported dataframe, then the merge
+            would be performed on ``year``.
+        merge_type: How the dataframes should be merged.
+            Values are ["left", "right", "outer", "inner", "cross"].
+            See :func:`pandas.DataFrame.merge`.
+        report_at_start: Whether the data in the less granular dataframe is reported
+            at the start of the time period.
+        kwargs : Additional arguments to pass to :func:`pandas.DataFrame.merge`.
+
+
+    Returns:
+        Merged contents of left and right input dataframes.
+
+    Raises:
+        ValueError: if ``left_date_col`` or ``right_date_col`` columns are missing from their
+            respective input dataframes.
+        ValueError: if any of the labels referenced in ``shared_merge_cols`` are missing from either
+            the left or right dataframes.
+
+    """
+
+    def separate_date_cols(df, date_col_name):
+        df[date_col_name] = pd.to_datetime(df[date_col_name])
+        df.loc[:, "year"] = df[date_col_name].dt.year
+        df.loc[:, "quarter"] = df[date_col_name].dt.quarter
+        df.loc[:, "month"] = df[date_col_name].dt.month
+        df.loc[:, "day"] = df[date_col_name].dt.day
+        df = df.drop(date_col_name, axis=1)
+        return df
+
+    right = right.copy()
+    left = left.copy()
+    right = separate_date_cols(right, right_date_col)
+    left = separate_date_cols(left, left_date_col)
+    merge_cols = temporal_merge_cols + shared_merge_cols
+    out = left.merge(right, on=merge_cols, how=merge_type, **kwargs)
+
+    # reconstruct the report_date column and do some column cleanup
```

Comment:
This reconstruction of `report_date` part isn't the cleanest. Basically if there is an x and y column for each of year, month, day, I took the larger value of the x and y sides as the more granular date (because the less granular date has the same value or 1 (january, first of month) if reported at the beginning of the period).